owasp-top10-2017-apps/a5/tictactoe/src/app.js

```diff
@@ -153,41 +154,48 @@ app.post('/login', async (req, res) => {
     const username = req.body.username
     const password = req.body.password
     const salt = await db.getPasswordSalt(username)
-    if (salt === undefined){
+    const type = await db.getType(username)
+    if (type === undefined) {
+        res
+            .status(400)
+            .json({ msg: "Type is undefined!" })
+    }
+    if (salt === undefined) {
         res
             .status(400)
-            .json({msg:"User doesn't exist or wrong password"})
+            .json({ msg: "User doesn't exist or wrong password" })
     }
     const bufferSalt = new Buffer(salt)
     const hashedPassword = crypto.hash(password, bufferSalt)
     const ok = await db.checkPassword(username, hashedPassword)
-    if (!ok){
+    if (!ok) {
         res
             .status(400)
-            .json({msg:"User doesn't exist or wrong password"})
+            .json({ msg: "User doesn't exist or wrong password" })
     }
 
-    const token = jwt.sign({ username }, process.env.SECRET, {
+    const token = jwt.sign({ type, username }, process.env.SECRET, {
         expiresIn: 3600
     });
-    
+
     res
         .cookie('tictacsession', token)
         .redirect('/game')
 });
 
-function verifyJWT(req, res, next){
+function verifyJWT(req, res, next) {
     var token = req.cookies.tictacsession
-    if (!token){
+    if (!token) {
         return res
-            .sendFile(path.join(__dirname+'/public/views/error.html'))
+            .sendFile(path.join(__dirname + '/public/views/error.html'))
     }
 
-    jwt.verify(token, process.env.SECRET, function(err, decoded){
-        if (err){
+    jwt.verify(token, process.env.SECRET, function (err, decoded) {
+        if (err) {
             return res
-                .sendFile(path.join(__dirname+'/public/views/error.html'))
+                .sendFile(path.join(__dirname + '/public/views/error.html'))
         }
+        if (decoded.username != req.body.username) return res.status(500).send({ auth: false, message: 'Failed to authenticate token.' });
```

Comment:
In this case, it is better to use a 400-class status, such as `401 Unauthorized`, since the credentials are invalid and the client can fix the error getting new credentials (or changing the username provided in request body)

Going forward, why do the client has to provide an `username` at all? She is already providing a JWT, and we can use that to determine the username instead of requiring it as a request parameter. What do you think, @thiagolotufo?